lls/cache.c

```diff
@@ -351,39 +350,80 @@ xmit_icmp_ping_reply(struct gatekeeper_if *iface, struct rte_mbuf *pkt)
 			RTE_ETHER_TYPE_IPV4);
 	}
 
-	/*
-	 * Set-up IPv4 header.
-	 *
-	 * According to RFC 792 page 13: to form an echo reply
-	 * message, the source and destination addresses are simply reversed,
-	 * the type code changed to 0, and the checksum recomputed.
-	 */
+	/* Set-up IPv4 header. */
 	icmp_ipv4 = (struct rte_ipv4_hdr *)pkt_out_skip_l2(iface, icmp_eth);
+	icmph = (struct rte_icmp_hdr *)ipv4_skip_exthdr(icmp_ipv4);
 
 	icmp_ipv4->time_to_live = IP_DEFTTL;
-	ip_addr_tmp = icmp_ipv4->src_addr;
-	icmp_ipv4->src_addr = icmp_ipv4->dst_addr;
-	icmp_ipv4->dst_addr = ip_addr_tmp;
+	if (icmph->icmp_type == ICMP_ECHO_REQUEST_TYPE) {
```

Comment:
This `if` statement is better rewritten as a `switch..case`. The default clause should report a serious error and drop the packet.